keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,497 @@
+
+#### Story 2
+
+As an extender of Kubernetes, my custom requested resources should be easily queryable by an administrator for scheduling without requiring Kubernetes code changes.
+
+### Story 3
+
+An administrator of the Kubernetes cluster should be able to see an aggregate representation of actual resource consumption metrics that is consistent with the resource request and scheduling model in order to measure resources that are imprecisely sized.
+
+### Notes/Constraints/Caveats (optional)
+
+The `kube-state-metrics` component of Kubernetes already exposes a lower fidelity model of resource requests from pods. However, because it is focused on exposing the attributes of the pod resource model rather than their calculated meaning, it captures an imprecise representation of the metrics Kubelet and the scheduler use to make decisions. This representation is not suitable for making complete decisions on *why* a resource is not schedulable, or to accurately capture how the scheduler views the resource model. In order to practically model these resources, a number of high cardinality metrics would need to be added and a fairly complicated calculation would have to be performed in the metrics aggregator that would subtly differ from the decisions made by the Kubelet or scheduler. Instead we recommend closely matching the implementation the scheduler uses to make decisions (documented and part of our API, but not explicitly modelled outside of code) to represent the key metrics for the resource model.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Expose new metrics
+
+The scheduler will expose an optional set of metrics series that capture the resources a pod is considered to be "consuming" from a scheduling perspective at the current time.
+
+The series reported will be consistent with `kube_node_status_allocatable` to simplify queries that compare used capacity and available capacity.
+
+* `kube_pod_resource_requests` contains series for the consumed requested resources for a pod at a given time.
+  * The value is the quantity of the limit in `spec.containers.resources.requests`
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `resource` - the name of the resource as described in `spec.containers.resources.requests`
+    * `units` - the units of the resource type as inferred from the resource name in `spec.containers.requests`, such as `bytes`, `cores`. Empty if this is a unitless resource.
```

Comment:
Echoing @brancz's comment, rather than including a units label, I think it would be more usable to include one timeseries per metrics type with units standardized which will allow for better/easier aggregation.

Reply:
The distinction here is between custom metrics and those in the standard model.  I called that out in one sentence in the next section, but will clarify here.

Standard model metrics would have a _type_unit name suffix.  Those without would be represented as labels in order to enable discoverability.

Reply:
Oops, let me clarify further.  There's a sentence above:

> The series reported will be consistent with `kube_node_status_allocatable` to simplify queries that compare used capacity and available capacity.

kube_node_status_allocatable uses units and resources, and these metrics are consistent with that to be comparable.  If we want to change the pattern, we should do it in both places.

The next section of metrics has a name suffix `_resource_units` form for standard units and resources.